### PR TITLE
Zhengwei/cpp autograder

### PR DESCRIPTION
**Pull request by zhengwei143**
- Mostly refactored code from the python implementation for the test cases
- Only works for simple cases

### CODE REVIEW

File: app/services/course/assessment/question/programming/programming_package_service.rb [hunk 1]
```diff
@@ -45,8 +45,9 @@ def init_language_package_service(params)
         Course::Assessment::Question::Programming::Python::PythonPackageService.new params
     elsif javascript_language
       raise NotImplementedError
-    else
-      raise NotImplementedError
+    elsif cpp_language
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/programming_package_service.rb]: At this rate, can we switch to a case statement instead, and remove all the methods? So it should look something like the one below. Also, you need to add back the case to catch all other languages. 

```ruby
def init_language_package_service(params)
  @language_package_service = begin
    case @language.class
    when Coursemology::Polyglot::Language::Python
      Course::Assessment::Question::Programming::Python::PythonPackageService.new params
    when ...
      ...
  end
end
```

File: app/services/course/assessment/question/programming/cpp/cpp_package_service.rb [hunk 2]
```diff
@@ -0,0 +1,248 @@
+# frozen_string_literal: true
+class Course::Assessment::Question::Programming::Cpp::CppPackageService < \
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: @zhengwei143 This service is a little long, would it be possible to break this down into smaller, independent pieces - any suggestions? 

Don't constrain yourself to the methods defined in `Course::Assessment::Question::Programming::LanguagePackageService`, as that can be refactored in the controller later on.

Comment (zhengwei143) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: I was thinking of waiting till I got the Java package done, since the Java package will have one on its own. Then extract the similarities and let them inherit from `Course::Assessment::Question::Programming::LanguagePackageService`. Then all the refactoring will be done in a separate PR after the Java package is complete. What do you think?

Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: I understand you had an offline conversation with @allenwq yesterday. Maybe you can work on that first, and keep this at the back of your head? Pushing first and then refactoring later seems ok, unless it makes testing for you too painful, then maybe splitting it up might be easier.

Comment (zhengwei143) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: yeah am currently looking into how to get the c/c++ autograder to work for multiple file submission. Should that be included in this PR as well?
As for the refactoring, it seems likely that there will be quite a bit of repeated code across the 3 packages so will be done eventually.

Comment (allenwq) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: @zhengwei143 we prefer to split the PR into smaller pieces, so that they can be reviewed and merged piece by piece, it will also be easier for you to only update a small piece.

I think you don't need to include the full multiple file implementation in this PR, but the design of the multiple files might affect the service you pushed here.

File: app/services/course/assessment/question/programming/cpp/cpp_package_service.rb [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    @meta = generate_meta(data_files_to_keep)
+
+    return nil if @meta == @old_meta
+
+    @attachment = generate_zip_file(data_files_to_keep)
+    FileUtils.remove_entry @tmp_dir if @tmp_dir.present?
+    @attachment
+  end
+
+  def default_meta
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: Can we shift this to the service superclass?

File: app/services/course/assessment/question/programming/cpp/cpp_package_service.rb [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+  AUTOGRADE_PRE_PATH = File.join(File.expand_path(File.dirname(__FILE__)),
+                                 'cpp_autograde_pre.cc').freeze
+
+  AUTOGRADE_POST_PATH = File.join(File.expand_path(File.dirname(__FILE__)),
+                                  'cpp_autograde_post.cc').freeze
+
+  MAKEFILE_PATH = File.join(File.expand_path(File.dirname(__FILE__)), 'cpp_makefile').freeze
+
+  def autograded?
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: I think we can move this to service superclass also. Make it a convention for services of all languages.

File: app/services/course/assessment/question/programming/cpp/cpp_package_service.rb [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+      end
+    end
+  end
+
+  def generate_non_autograded_meta(template_files)
+    meta = default_meta
+
+    return meta if template_files.blank?
+
+    # For python editor, there is only a single submission template file.
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: I think you need to change this to cpp

File: app/services/course/assessment/question/programming/cpp/cpp_package_service.rb [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    meta[:data_files] = get_data_files_meta(data_files_to_keep, new_data_files)
+
+    [:public, :private, :evaluation].each do |test_type|
+      meta[:test_cases][test_type] = @test_params[:test_cases][test_type] || []
+    end
+
+    meta.as_json
+  end
+
+  def test_params(params)
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: I think you can shift this to the superclass too. The subclasses have the ability to overwrite this if they need a slightly different param.

File: app/services/course/assessment/question/programming/cpp/cpp_package_service.rb [hunk 7]
```diff
@@ -0,0 +1,219 @@
+# frozen_string_literal: true
+class Course::Assessment::Question::Programming::Cpp::CppPackageService < \
+  Course::Assessment::Question::Programming::LanguagePackageService
+  AUTOGRADE_INCLUDE_PATH = File.join(File.expand_path(File.dirname(__FILE__)),
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: If this is repeated a lot, can also refactor this and place it somewhere to be included.

Comment (zhengwei143) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: These specify the path to the files used in each language package service, which also depend on the package structure and filenames, which are different for each service. Will be better to keep it within the class itself.

Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: As in `AUTOGRADE_INCLUDE_PATH = current_path_for('cpp_autograde_include.cc')`.

Comment (zhengwei143) [inline app/services/course/assessment/question/programming/cpp/cpp_package_service.rb]: Oh i see, will move it to the LanguagePackageService super class then

File: app/services/course/assessment/question/programming/cpp/cpp_autograde_pre.cc [hunk 8]
```diff
@@ -0,0 +1,132 @@
+// Variant type that encompasses all the types that expect_equal() supports
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/cpp/cpp_autograde_pre.cc]: Could you also just add in the 'why' for this part, on top of the 'what'? At least this explains the issue, and how the overloading here overcomes that. Might be helpful for someone who wishes to 'refactor' this in the future, or extend it.

File: app/services/course/assessment/question/programming/language_package_service.rb [hunk 9]
```diff
@@ -60,14 +68,37 @@ def extract_meta(attachment, template_files) # rubocop:disable UnusedMethodArgum
     raise NotImplementedError, 'You must implement this'
   end
 
+  # Retrieves the absolute path of the file specified
+  #
+  # @param [String] curr_dir The current directory of the specific language package service
+  # @param [String] filename The filename of the file to get the path of
+  def self.get_file_path(curr_dir, filename)
+    File.join(File.expand_path(File.dirname(__FILE__)), curr_dir, filename).freeze
```
Comment (weiqingtoh) [inline app/services/course/assessment/question/programming/language_package_service.rb]: If I am not wrong, I think there is no need to specify `curr_dir`, replace it with `self` and it should work. The self in this case evaluates to the context of the caller, which would be the specific language service. 

Can try out and verify?

**Comment by weiqingtoh** [comment]
After addressing those issues, I think we can proceed with this for now. @allenwq will leave you to review and merge 👍 

When doing the Java service, I'd suggest to refactor the service a little more. I think there will be some common functionality, like checking whether attachments have been overwritten etc.

File: app/services/course/assessment/question/programming/cpp/cpp_autograde_pre.cc (hunk 8 above)
Comment (zhengwei143) [inline app/services/course/assessment/question/programming/cpp/cpp_autograde_pre.cc]: I decided to remove the use of the boost library. Initially required it but after changing the structure, i realized that it would not be needed anymore. Would be better also, since it makes things less complicated ><

File: app/services/course/assessment/question/programming/language_package_service.rb (hunk 9 above)
Comment (zhengwei143) [inline app/services/course/assessment/question/programming/language_package_service.rb]: Doesn't work, i've refactored it as a private method in each `*_package_service` instead, and moved the variables into the function (since they are only being used once).

**Comment by allenwq** [comment]
LGTM, @fonglh  ?
